connectors/lang.js.php

```diff
@@ -5,68 +5,115 @@
  * @var modX $modx
  * @package modx
  * @subpackage lexicon
+ *
  */
+
+function esc($s) {
+    return strtr($s, array('\\' => '\\\\', "'" => "\\'", '"' => '\\"', "\r" => '\\r', "\n" => '\\n', '</' => '<\/'));
+}
+
+$time = microtime(true);
+
 define('MODX_CONNECTOR_INCLUDED', 1);
 
-ob_start();
 require_once dirname(__FILE__).'/index.php';
-ob_clean();
 
-if (!empty($_REQUEST['topic'])) {
-    $topics = explode(',',$_REQUEST['topic']);
-    foreach($topics as $topic) $modx->lexicon->load($topic);
-}
+/* we compute the whole dictionary for the used language at once and use the cache */
+/* can be kept infinite until cache is refreshed */
+$cacheOptions = array(
+    xPDO::OPT_CACHE_KEY => 'lexicon-js',
+    xPDO::OPT_CACHE_EXPIRES => 0
+);
+$lang = $modx->getOption('manager_language','en');
+$cachedOutput = $modx->cacheManager->get('lang.js_'.$lang, $cacheOptions);
 
-$entries = $modx->lexicon->fetch();
-echo '
-MODx.lang = {';
-$s = '';
-while (list($k,$v) = each ($entries)) {
-    $s .= "'$k': ".'"'.esc($v).'",';
-}
-$s = trim($s,',');
-echo $s.'
-};
-var _ = function(s,v) {
-    if (v != null && typeof(v) == "object") {
-        var t = ""+MODx.lang[s];
-        for (var k in v) {
-            t = t.replace("[[+"+k+"]]",v[k]);
-        }
-        return t;
-    } else return MODx.lang[s];
-}';
+/* create output if nothing is cached */
+if (is_null($cachedOutput)) {
 
-function esc($s) {
-    return strtr($s,array('\\'=>'\\\\',"'"=>"\\'",'"'=>'\\"',"\r"=>'\\r',"\n"=>'\\n','</'=>'<\/'));
-}
+    /* get the prefixed topic names for each namespace */
+    $tocs = array();
+    foreach ($modx->getCollection('modNamespace') as $n) {
+        $tocs = array_merge($tocs,
+            preg_filter(
+                '/^/', $n->name.':',
+                $modx->lexicon->getTopicList( $lang, $n->name )
+            )
+        );
+    }
 
-/* gather output from buffer */
-$output = ob_get_contents();
-ob_end_clean();
+    /* load all topics from all namespaces and fetch the entries */
+    call_user_func_array( array(&$modx->lexicon,'load'), $tocs);
+    $entries = $modx->lexicon->fetch();
 
+    $buf = 'MODx.lang = {';
+    $s = '';
+    while (list($k, $v) = each($entries)) {
+        $s .= "'$k': " . '"' . esc($v) . '",';
+    }
+    $s = trim($s, ',');
+    $buf .= $s;
+
+    // use compressed js here - saves some bytes :-)
```

Comment:
Kinda ironic that this PR increases the size of the lang.js ten-fold, but we need to use a compressed js snippet to save bytes?